### ISSUE TITLE
Return a 503 not a 404 when all instances down

### ISSUE THREAD

**Issue by mikebl71MC** [cloudfoundry/routing-release#250]
## Issue

When all instances of an application are not available, clients get 404 Not Found instead of 503 Service Unavailable.

## Affected Versions

PCF 2.12

## Context

If a route is registered but all instances of an application are currently down then GoRouter currently returns 404 Not Found. It would 
be better if it returned 503 Service Unavailable.

It looks like this issue was already discussed in the past:
https://github.com/cloudfoundry/gorouter/pull/251
https://github.com/cloudfoundry/gorouter/issues/135

As far as I understand, you don't object to this change in principal but this would "involve some substantial changes to how routes are registered".
How can we re-open this issue and get the changes implemented?

This behavior currently causes problems for our customers. Most of our APIs are RESTful. So when a customer sends us, e.g., GET /resources/12345 and gets back 404 Not Found, they assume that this rest resource does not exist and trigger their business logic.  They do not expect that this situation is also possible when our application is temporarily down.

## Steps to Reproduce

Deploy a java application and register a route.
Stop the application. (Any other way of killing the application produces the same result.)
Client sends an http request to an application's endpoint.

## Expected result

Client should receive 503 Service Unavailable.

## Current result

Client receives 404 Not Found.
(To be precise, in a few seconds after the app goes down, the client receives 502 Bad Gateway and 503 Service Unavailable, but later it always receives 404 Not Found).

**Comment by plowin** [comment]
Hi @mikebl71MC, 
thanks for raising your request!

I followed the well-documented investigations and findings @ameowlia shared in https://github.com/cloudfoundry/gorouter/pull/251#issuecomment-657638000.
Going through the steps, I could verify that the HTTP header "x-cf-routererror" is set for euch of the different status codes (502, 503, 404):
* 502 Bad Gateway: Registered endpoint failed to handle the request.
  < x-cf-routererror: endpoint_failure (dial tcp 10.0.73.2:61074: connect: connection refused)
* 503 Service Unavailable: Requested route ('my.app.com') has no available endpoints.
  < x-cf-routererror: no_endpoints
* 404 Not Found: Requested route ('my.app.com') does not exist.
  < x-cf-routererror: unknown_route

The x-cf-routererror header is not set if the backend app is returning a 404. Even though I fully agree that the current behaviour is unfortunate, couldn't it be a workaround to distinguish on customer-side based on this header?



Apart from that, some thoughts on the comment of @ameowlia in 2020:
> ... would require the gorouter to get routes from the CC API. This would be a very large change to the system.

Currently, the gorouter only has a runtime-dependency to the NATS messaging system (publish-subscribe). This lightweight architecture is working very well and is fully decoupled from the CC. I am not sure if we could keep our valuable status quo while introducing such a runtime dependency to the CC.

I will also share this request within our team. Possibly, other ideas might be raised.

Best regards,
Patrick

**Comment by n-shaw** [comment]
Could a customer create a [routing service](https://docs.cloudfoundry.org/services/route-services.html) in front of any services where they would prefer a `503` instead of a `404`?  Seems like a routing service would be able to see the `x-cf-routererror` response header and then change the response status to a `503`.

It may not be ideal, but if you are leveraging the HTTP status as a means of communicating data (confirmation that a resource does not exist), I can see how you might need to take those extra steps.

**Comment by domdom82** [comment]
I think the 404 gets returned eventually, because the pruning algorithm keeps removing routes with no endpoints here:
https://github.com/cloudfoundry/gorouter/blob/main/registry/registry.go#L322

Since gorouter is stateless, it cannot know if a backend pool is empty because the app crashed and is about to recover or if the app is just shut down. So I guess some kind of clean up is needed either way to remove "empty routes" with no endpoints.

However, in practice what we see is apps crash and eventually recover, so their responses turn from 502 -> 503 -> 404 and back to 200 again after some time.

@ameowlia What if we gave the [Endpoint Pool](https://github.com/cloudfoundry/gorouter/blob/main/route/pool.go#L141) also an attribute `UpdatedAt` like [Endpoint](https://github.com/cloudfoundry/gorouter/blob/main/route/pool.go#L77) that gets refreshed whenever an endpoint is added or removed? We could then have another timeout that allows an empty pool to stay around and evade pruning in order to give a recovering app time to register its routes again. This would allow us to widen the time range where a 503 is returned and shrink the 404 range.

**Comment by ameowlia** [comment]
@domdom82 what a creative solution! Much better than getting all routes from the cloud controller. 

If someone submitted that PR I would happily merge it.

**Comment by plowin** [comment]
Hi, thanks for the feedback, @ameowlia.
We will try to work on this during Q1/early Q2 2022!

**Comment by mikebl71MC** [comment]
Thank you for the response and for working on a solution.
a) could you please make the default value for the above timeout reasonably high, at least 30 sec, maybe even more. So that the app has a good chance to recover. Or, if longer problems, then the infrastructure has enough time to spot the problem and move traffic to another region.
b) just to confirm, would the above solution work regardless of whether the app has crashed or it is stopped manually?
Sometimes we might need to stop the app eg for database maintenance. And would be good if the GoRouter behaviour was the same in both cases.

**Comment by domdom82** [comment]
@mikebl71MC Thanks for your feedback. We will play around with timeouts to find a reasonable default value. We used to have 20 secs for routes refresh and I think pruning started after 2 mins if not refreshed. So this could be a good starting point.

Regarding b) I don't think it makes a difference. When you stop an app for maintenance, you should see the same behavior.

**Comment by domdom82** [comment]
draft PR: https://github.com/cloudfoundry/gorouter/pull/314

**Comment by domdom82** [comment]
Hi @ameowlia I think the changes are ready for review now.
tldr:
- There is now a `empty_pool_timeout` property on gorouter
- The default value is 0 because we didn't want to change default behavior and act defensively (@mikebl71MC suggested at least 30s default, which we can also do)
- The property is used in conjunction with the `empty_pool_response_code_503` flag
- If both are set, the `empty_pool_timeout` controls, how long gorouter will return 503 instead of 404 once there are no more routes in the pool
- This affects failed endpoints, crashed apps and stopped apps 

PRs: 
https://github.com/cloudfoundry/routing-release/pull/268
https://github.com/cloudfoundry/gorouter/pull/314

**Comment by ameowlia** [comment]
Hi all,

🎉 First off a big thank you to @mikebl71MC for raising this issue, thank you @kecirlotfi for submitting the PRs, and thank you @domdom82 for your great solution and for your reviews. 

I agree with you @domdom82. I think the default value should be 0. 

I merged the PRs and can't wait to release!
